File: ch1/1-20.c
#include <stdio.h>

#define MAXLINE 1000
#define TABLEN 4

int get_line(char s[], int lim);
void detab(char s[], int len);

int main(void)
{
    char line[MAXLINE];
    int len;
    while ((len = get_line(line, MAXLINE)) > 0) {
        detab(line, len);
        printf("%s\n", line);
    }
    return 0;
}

int get_line(char s[], int lim)
{
    int c, i;
    for (i = 0; i < lim - 1 && (c = getchar()) != EOF && c != '\n'; i++) {
        s[i] = c;
    }
    if (c == '\n') {
        s[i] = c;
        ++i;
    }
    s[i] = '\0';

    return i;
}

void detab(char s[], int len)
{
    int found_times = 0;
    for (int i = 0; i < len; i++) {
        if (s[i] == '\t') {
            found_times++;
            s[i] = ' ';
            for (int j = len-1; j >= i+1; j--) {
                s[j+(TABLEN-1)] = s[j];
            }
            for (int k = 1; k < TABLEN; k++) {
                s[i+k] = ' ';
            }
        }
    }
    len += (TABLEN*found_times);
    s[len] = '\0';
}

File: ch1/1-20.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const maxLine = 1000
const tabLen = 4

func main() {
	for {
		line, l := getLine(maxLine)
		detab(line, l)
		fmt.Printf("%s\n", string(line))
	}
}

func detab(s []byte, l int) {
	for i := 0; i < l; i++ {
		if rune(s[i]) == '\t' {
			s[i] = byte(' ')
			for j := l - 1; j >= i+1; j-- {
				s[j+(tabLen-1)] = s[j]
			}
			for k := 1; k < tabLen; k++ {
				s[i+k] = byte(' ')
			}
		}
	}
}

func getLine(lim int) ([]byte, int) {
	var c byte
	var i int
	s := make([]byte, lim)
	r := bufio.NewReader(os.Stdin)
	for i = 0; i < lim-1; i++ {
		c, err := r.ReadByte()
		if rune(c) == '\n' || err != nil {
			break
		}
		s[i] = c
	}
	if rune(c) == '\n' {
		s[i] = c
		i++
	}
	return s, i
}
